lib/patches/thread_safe/thread_safe.rb

```diff
@@ -0,0 +1,180 @@
+#
+# 6. Trademarks. This License does not grant permission to use the trade names, trademarks,
+# service marks, or product names of the Licensor, except as required for reasonable and
+# customary use in describing the origin of the Work and reproducing the content of the NOTICE
+# file.
+#
+# 7. Disclaimer of Warranty. Unless required by applicable law or agreed to in writing,
+# Licensor provides the Work (and each Contributor provides its Contributions) on an "AS IS"
+# BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied, including,
+# without limitation, any warranties or conditions of TITLE, NON-INFRINGEMENT,
+# MERCHANTABILITY, or FITNESS FOR A PARTICULAR PURPOSE. You are solely responsible for
+# determining the appropriateness of using or redistributing the Work and assume any risks
+# associated with Your exercise of permissions under this License.
+#
+# 8. Limitation of Liability. In no event and under no legal theory, whether in tort
+# (including negligence), contract, or otherwise, unless required by applicable law (such as
+# deliberate and grossly negligent acts) or agreed to in writing, shall any Contributor be
+# liable to You for damages, including any direct, indirect, special, incidental, or
+# consequential damages of any character arising as a result of this License or out of the use
+# or inability to use the Work (including but not limited to damages for loss of goodwill,
+# work stoppage, computer failure or malfunction, or any and all other commercial damages or
+# losses), even if such Contributor has been advised of the possibility of such damages.
+#
+# 9. Accepting Warranty or Additional Liability. While redistributing the Work or Derivative
+# Works thereof, You may choose to offer, and charge a fee for, acceptance of support,
+# warranty, indemnity, or other liability obligations and/or rights consistent with this
+# License. However, in accepting such obligations, You may act only on Your own behalf and on
+# Your sole responsibility, not on behalf of any other Contributor, and only if You agree to
+# indemnify, defend, and hold each Contributor harmless for any liability incurred by, or
+# claims asserted against, such Contributor by reason of your accepting any such warranty or
+# additional liability.
+#
+# END OF TERMS AND CONDITIONS
+
+# Adapt code from the thread_safe gem intended to support thread-safe collections in Rubinius. TruffleRuby doesn't
+# honor `allocate' authored in Ruby like Rubinius does, so this code has been updated to use `initialize' instead.
+module ThreadSafe
+  require 'monitor'
+
+  class Hash < ::Hash; end
+  class Array < ::Array; end
+
+  [Hash, Array].each do |klass|
+    klass.class_eval do
+      private
+      def _mon_initialize
+        @_monitor = Monitor.new unless @_monitor # avoid double initialisation
+      end
+
+      def initialize(*args)
+        obj = super
+        obj.send(:_mon_initialize)
+        obj
```

Comment:
There is no need to return the object in `#initialize`, and no guarantee `super` returns `self` (in general).
So this should be just
```ruby
  super
  _mon_initialize
```